File: Lua/SelectionSort.lua
--根据索引交换
function Swap(arr,index1,index2) 
  local temp = arr[index1];
  arr[index1] = arr[index2];
  arr[index2] = temp;
end

--找到最小的Index
function FindMinIndex(arr,startIndex)
  local val = arr[startIndex];
  local idx = startIndex;
  for i=startIndex+1,#arr  do
    if(val > arr[i]) then
      idx = i;
      val = arr[i];
    end
  end
  return idx;
end

function SelectionSort(arr) 
  for i=1,#arr do
    local minIndex = FindMinIndex(arr,i);
    Swap(arr,i,minIndex);
  end
end

local arr = {88,5,2,1,66};
SelectionSort(arr);
for i=1,#arr do
  print(arr[i]);
end

File: C#/Algorithm/SelectionSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Algorithm
{
    /// <summary>
    /// 选择排序，每次循环找出最大/最小的值放置到相应位置
    /// </summary>
    public class SelectionSort
    {
        private static int FindMin(int[] arr, int startIndex)
        {
            int minVal = arr[startIndex];
            int minIdx = startIndex;

            for (int i = startIndex + 1; i < arr.Length; i++)
            {
                if (minVal > arr[i])
                {
                    minVal = arr[i];
                    minIdx = i;
                }
            }

            return minIdx;
        }

        public static void Sort(int[] arr)
        {
            for (int i = 0; i < arr.Length; i++)
            {
                int minIndex = FindMin(arr, i);
                Util.Swap(arr, minIndex, i);
            }
        }
    }
}
